### PR TITLE
feat: add init prompts for Gen2

### PR DESCRIPTION
**Pull request by 0618**
#### Description of changes

This PR is to add some prompts before a user init a new Gen1 project.

- If user decide to not continue with Gen1, exit the process.
Screenshot 2024-07-10 at 16 21 28

- If user decide to continue with Gen1, ask for the reason

Screenshot 2024-07-10 at 16 22 32

- Add the reason to project-config.json
Screenshot 2024-07-09 at 17 53 18

- Add to stack metadata

Screenshot 2024-07-10 at 16 31 26

- Only show the prompts in a new project. 
- The change doesn't affect headless mode.

#### Description of how you validated changes

- tested all the scenarios above on my local machine.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-cli/src/init-steps/preInitMigrationWarning.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  printer.warn(
+    `AWS Amplify recommends using Amplify Gen 2 for new projects. Learn how to get started at https://docs.amplify.aws/react/start/quickstart/`,
+  );
+
+  const continueWithGen1 = await prompter.confirmContinue('Do you want to continue with Amplify Gen 1?');
+
+  if (!continueWithGen1) {
+    printer.info('Exiting the Amplify Gen1 project creation flow. 👋');
+    printer.blankLine();
+    printer.info('To create a new Amplify project using Amplify Gen 2 🫱 https://docs.amplify.aws/');
```
Comment (Amplifiyer) [inline packages/amplify-cli/src/init-steps/preInitMigrationWarning.ts:18]:
```suggestion
    printer.info('To create a new Amplify Gen2 project, 🫱 https://docs.amplify.aws/');
```

File: packages/amplify-cli/src/commands/init.ts [hunk 2]
```diff
@@ -10,6 +10,7 @@ import { initProviders } from '../init-steps/s2-initProviders';
 import { scaffoldProjectHeadless } from '../init-steps/s8-scaffoldHeadless';
 import { onHeadlessSuccess, onSuccess } from '../init-steps/s9-onSuccess';
 import { checkForNestedProject } from './helpers/projectUtils';
+import { migrationWarning } from '../init-steps/preInitMigrationWarning';
```
Comment (Amplifiyer) [inline packages/amplify-cli/src/commands/init.ts:13]: It's not migration related warning. Perhaps `gen2Recommendation`?

**Review by Amplifiyer — COMMENTED** [review]
Should there be unit tests for this change?

File: packages/amplify-cli/src/commands/init.ts (hunk 2 above)
Comment (0618) [reply packages/amplify-cli/src/commands/init.ts]: Good idea. Done [49a9f01](https://github.com/aws-amplify/amplify-cli/pull/13849/commits/49a9f01f2e70fd3c9b41b336b643e38f2e47d55f)

File: packages/amplify-cli/src/init-steps/gen2Recommendation.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import { $TSContext } from '@aws-amplify/amplify-cli-core';
+import { printer, prompter } from '@aws-amplify/amplify-prompts';
+import { isNewProject } from './s0-analyzeProject';
+
+export const gen2Recommendation = async (context: $TSContext): Promise<$TSContext> => {
+  if (!isNewProject(context)) {
+    return context;
+  }
+  printer.warn(
+    `AWS Amplify recommends using Amplify Gen 2 for new projects. Learn how to get started at https://docs.amplify.aws/react/start/quickstart/`,
```
Comment (edwardfoyle) [inline packages/amplify-cli/src/init-steps/gen2Recommendation.ts:10]: Unless this wording is coming from legal/product/somewhere we don't control, it sounds awkward to say "AWS Amplify recommends...". Instead I would just say
```suggestion
`Amplify Gen 2 is recommended for new projects. Learn how to get started at https://docs.amplify.aws/react/start/quickstart/`,
```

File: packages/amplify-cli/src/commands/init.ts [hunk 4]
```diff
@@ -22,7 +23,7 @@ const constructExeInfo = (context: $TSContext): void => {
 const runStrategy = (quickstart: boolean) =>
   quickstart
     ? [preInitSetup, analyzeProjectHeadless, scaffoldProjectHeadless, onHeadlessSuccess]
-    : [preInitSetup, analyzeProject, initFrontend, initProviders, onSuccess, postInitSetup];
+    : [migrationWarning, preInitSetup, analyzeProject, initFrontend, initProviders, onSuccess, postInitSetup];
```
[the remark this answers is not in the document]
Comment (0618) [reply packages/amplify-cli/src/commands/init.ts]: refactored 
0d3a057

File: packages/amplify-cli/src/init-steps/gen2Recommendation.ts (hunk 3 above)
Comment (0618) [reply packages/amplify-cli/src/init-steps/gen2Recommendation.ts]: Yes, the wording is from the ticket 🤦‍♀️

File: packages/amplify-cli/src/init-steps/preInitSetup.ts [hunk 5]
```diff
@@ -22,6 +35,38 @@ export const preInitSetup = async (context: $TSContext): Promise<$TSContext> =>
...
+  }
+
+  const whyContinueWithGen1 = await prompter.pick('Why would you like to use Amplify Gen 1?', [
+    'I am a current Gen 1 user',
+    'Gen 2 is missing features I need from Gen 1',
+    'I find the Gen 1 CLI easier to use',
+    'Prefer not to answer',
+  ]);
+
+  context.exeInfo.projectConfig = {
```
Comment (edwardfoyle) [inline packages/amplify-cli/src/init-steps/preInitSetup.ts:63]: does `context.exeInfo.projectConfig` already have any content at this point? If so, this line would erase that

File: packages/amplify-cli/src/commands/init.ts [hunk 6]
```diff
@@ -18,11 +18,12 @@ const constructExeInfo = (context: $TSContext): void => {
   };
 };
 
+const isHeadless = true;
```
Comment (edwardfoyle) [inline packages/amplify-cli/src/commands/init.ts:21]: I would name this input `recommendGen2: true|false`

File: packages/amplify-cli/src/init-steps/preInitSetup.ts [hunk 7]
```diff
@@ -3,11 +3,24 @@ import { execSync } from 'child_process';
 import * as fs from 'fs-extra';
 import * as url from 'url';
 import { generateLocalEnvInfoFile } from './s9-onSuccess';
+import { printer, prompter } from '@aws-amplify/amplify-prompts';
+import { isNewProject } from './s0-analyzeProject';
+
+export const preInitSetup = (isHeadless) => {
```
Comment (edwardfoyle) [inline packages/amplify-cli/src/init-steps/preInitSetup.ts:9]: Since this is a higher-order function now, it should probably be named `getPreInitSetup`

File: packages/amplify-cli/src/init-steps/gen2Recommendation.ts (hunk 3 above)
Comment (swaminator) [reply packages/amplify-cli/src/init-steps/gen2Recommendation.ts]: Agree with the recommendation

File: packages/amplify-cli/src/init-steps/preInitSetup.ts (hunk 5 above)
Comment (0618) [reply packages/amplify-cli/src/init-steps/preInitSetup.ts:63]: It does not yet, because it's the `preInitSetup`. 

In the same PR in the s0-analyzeProject.ts file, it was where `projectConfig` was first generated, so `...context.exeInfo.projectConfig,` is added to prevent overriden.

File: packages/amplify-cli/src/commands/init.ts (hunk 6 above)
Comment (0618) [reply packages/amplify-cli/src/commands/init.ts:21]: hmmmm.... that makes sense for this PR, (might be unlikely) but what if we add something else to `preInitSetup` not Gen2 related?

File: packages/amplify-cli/src/init-steps/preInitSetup.ts (hunk 7 above)
Comment (0618) [reply packages/amplify-cli/src/init-steps/preInitSetup.ts:9]: Good point. Done!

File: packages/amplify-cli/src/init-steps/preInitSetup.ts (hunk 5 above)
Comment (0618) [reply packages/amplify-cli/src/init-steps/preInitSetup.ts:63]: Yes. Before this PR, it was first added in s0-analyzeProject.ts line 154 in this PR.

File: packages/amplify-cli/src/init-steps/preInitSetup.ts (hunk 7 above)
Comment (0618) [reply packages/amplify-cli/src/init-steps/preInitSetup.ts]: [08e896a](https://github.com/aws-amplify/amplify-cli/pull/13849/commits/08e896acd6a3562f1c03fce975b33c92f7fc9891)

File: packages/amplify-cli/src/init-steps/gen2Recommendation.ts (hunk 3 above)
Comment (0618) [reply packages/amplify-cli/src/init-steps/gen2Recommendation.ts]: Done [22cff01](https://github.com/aws-amplify/amplify-cli/pull/13849/commits/22cff01e5e878bdf59bbb86ef5000c3fb21fef7e)

File: .circleci/local_publish_helpers_codebuild.sh [hunk 8]
```diff
@@ -323,7 +319,7 @@ function checkPackageVersionsInLocalNpmRegistry {
 
     if [[ $cli_internal_version != $cli_version ]]; then
         echo "Versions did not match."
-        echo "Manual fix: add a proper conventional commit that touches the amplify-cli-npm package to correct its version bump. For example https://github.com/aws-amplify/amplify-cli/commit/6f14792d1db424aa428ec4836fed7d6dd5cccfd0"
+        echo "Manual fix: add a proper conventional commit that touches the amplify-cli-npm package to correct its version bump. For example https://github.com/aws-amplify/amplify-cli/pull/13759/commits/15dcd96feae925ff26ca51abfb4a0477890af745"
```
Comment (awsluja) [inline .circleci/local_publish_helpers_codebuild.sh:322]: you will need to manually kick of PR workflow since this touches the /.circleci folder

File: packages/amplify-cli/src/commands/init.ts (hunk 6 above)
Comment (edwardfoyle) [reply packages/amplify-cli/src/commands/init.ts:21]: If we're worried about that, we can pass in an options object `{recommendGen2: boolean}`. However, since this is an internal function, that could always be a future refactor if there are more options later on

File: packages/amplify-cli/src/init-steps/preInitSetup.ts (hunk 5 above)
Comment (edwardfoyle) [reply packages/amplify-cli/src/init-steps/preInitSetup.ts:63]: Looks like that file might blow this away then: https://github.com/aws-amplify/amplify-cli/blob/2570d0c448b63e6cad419757030fd9f26eb97557/packages/amplify-cli/src/init-steps/s0-analyzeProject.ts#L154

Where is this value being consumed? Do we have tests around that?

File: packages/amplify-cli-npm/index.ts [hunk 9]
```diff
@@ -16,4 +16,5 @@ export const install = async (): Promise<void> => {
   return binary.install();
 };
 
-// force version bump to 12.12.0
+//
+// force version bump to 12.12.5
```
[the remark this answers is not in the document]
Comment (0618) [reply packages/amplify-cli-npm/index.ts]: oops...@awsluja also mentioned that. I forgot to change it.

fixed it cac0e9d

File: packages/amplify-cli/src/init-steps/preInitSetup.ts (hunk 5 above)
Comment (0618) [reply packages/amplify-cli/src/init-steps/preInitSetup.ts:63]: It won't be blown away because it's already taken cared in this PR

https://github.com/aws-amplify/amplify-cli/pull/13849/files#diff-472b813f47ea0bbcba3522fb7e9f5ae4633dc0f14d44b8a074370baa67e5d961R155

We don't have integ test for the stack metadata yet...

File: packages/amplify-cli/src/commands/init.ts (hunk 6 above)
Comment (0618) [reply packages/amplify-cli/src/commands/init.ts]: Done dfba32eb5086bd4918c3daebe949ca439c49f543

File: packages/amplify-cli/src/init-steps/preInitSetup.ts [hunk 10]
```diff
@@ -22,6 +35,38 @@ export const preInitSetup = async (context: $TSContext): Promise<$TSContext> =>
   return context;
 };
 
+/**
+ * recommend using Gen 2 or continue with Gen 1.
+ * ask for why they are using Gen 1 and store the answer in project-config
+ */
+export const gen2Recommendation = async (context: $TSContext): Promise<$TSContext> => {
```
Comment (Amplifiyer) [inline packages/amplify-cli/src/init-steps/preInitSetup.ts:42]: I think it doesn't need to be exported anymore?

File: .circleci/local_publish_helpers_codebuild.sh [hunk 11]
```diff
@@ -214,12 +213,11 @@ function retry {
...
     FAILED_TEST_REGEX_FILE="./amplify-e2e-reports/amplify-e2e-failed-test.txt"
     rm -f $FAILED_TEST_REGEX_FILE
-    until [ $n -ge $MAX_ATTEMPTS ]
-    do
+    until [ $n -ge $MAX_ATTEMPTS ]; do
         echo "Attempting $@ with max retries $MAX_ATTEMPTS"
         setAwsAccountCredentials
         "$@" && break
-        n=$[$n+1]
+        n=$(($n + 1))
```
Comment (awsluja) [inline .circleci/local_publish_helpers_codebuild.sh:220]: are you using some custom .sh formatter ?

File: packages/amplify-cli/src/init-steps/preInitSetup.ts [hunk 12]
```diff
@@ -3,11 +3,24 @@ import { execSync } from 'child_process';
...
+      await preInitSetupBasic(context);
+    };
+  }
+};
 
 /**
  * Executes before init
  */
-export const preInitSetup = async (context: $TSContext): Promise<$TSContext> => {
+export const preInitSetupBasic = async (context: $TSContext): Promise<$TSContext> => {
```
Comment (Amplifiyer) [inline packages/amplify-cli/src/init-steps/preInitSetup.ts:23]: I think it's fine to leave the name as-is. And possible it doesn't need to be exported anymore.
```suggestion
export const preInitSetup = async (context: $TSContext): Promise<$TSContext> => {
```

File: packages/amplify-cli/src/init-steps/s0-analyzeProject.ts [hunk 13]
```diff
@@ -152,6 +152,7 @@ export const analyzeProject = async (context: $TSContext): Promise<$TSContext> =
 const setProjectConfig = (context: $TSContext, projectName: string): void => {
   context.exeInfo.isNewProject = isNewProject(context);
   context.exeInfo.projectConfig = {
+    ...context.exeInfo.projectConfig,
```
Comment (awsluja) [inline packages/amplify-cli/src/init-steps/s0-analyzeProject.ts:155]: why was this added? was there a reason we didn't want to copy all properties from the config before?

**Review by Amplifiyer — APPROVED** [review]
Looks good overall, left some non-blocking comments from my end.

**Comment by awsluja** [comment]
can you run e2e?

File: packages/amplify-cli/src/init-steps/s0-analyzeProject.ts (hunk 13 above)
Comment (0618) [reply packages/amplify-cli/src/init-steps/s0-analyzeProject.ts:155]: This was added because we *do* want to carry over all properties from the config before.

File: .circleci/local_publish_helpers_codebuild.sh (hunk 8 above)
Comment (0618) [reply .circleci/local_publish_helpers_codebuild.sh:326]: reverted the format fba0211cc9bb958db529c2ec4c9ae4f07abf46f4

File: packages/amplify-cli/src/init-steps/preInitSetup.ts (hunk 12 above)
Comment (0618) [reply packages/amplify-cli/src/init-steps/preInitSetup.ts]: Done 1e8e25ba3bf8efb06a0096f000c74f7d1dc4caff

File: packages/amplify-cli/src/init-steps/preInitSetup.ts (hunk 10 above)
Comment (0618) [reply packages/amplify-cli/src/init-steps/preInitSetup.ts:42]: We need to export it for the unit test

Comment (0618) [reply packages/amplify-cli/src/init-steps/preInitSetup.ts:42]: reverted the export 
1e8e25b

File: .circleci/local_publish_helpers_codebuild.sh (hunk 11 above)
Comment (0618) [reply .circleci/local_publish_helpers_codebuild.sh]: reverted.

File: packages/amplify-provider-awscloudformation/src/template-description-utils.ts [hunk 14]
```diff
@@ -83,12 +83,15 @@ export function getDefaultTemplateDescription(context: $TSContext, category: str
   // get stack type "stackType"
   const stackTypeDescription = service ? `${category}-${service}` : category;
 
+  // get the reason why not switching to Gen2
+  const { whyContinueWithGen1 } = context.exeInfo.projectConfig;
```
Comment (awsluja) [inline packages/amplify-provider-awscloudformation/src/template-description-utils.ts:87]: do you know if this is defined yet?

Comment (0618) [reply packages/amplify-provider-awscloudformation/src/template-description-utils.ts]: It'd be defined in normal cases, but undefined when running `amplify export` if the project is created with `--yes` at `amplify init`.

Comment (0618) [reply packages/amplify-provider-awscloudformation/src/template-description-utils.ts]: Fixed in [faa0859](https://github.com/aws-amplify/amplify-cli/pull/13849/commits/faa0859f81be5ef816e8a7f6e40bba88477390d0)

File: packages/amplify-cli/src/init-steps/preInitSetup.ts [hunk 15]
```diff
@@ -3,6 +3,19 @@ import { execSync } from 'child_process';
...
+
+export const getPreInitSetup = (recommendGen2: boolean) => {
+  if (recommendGen2) {
+    return preInitSetup;
+  } else {
+    return async (context) => {
+      await gen2Recommendation(context);
+      await preInitSetup(context);
+    };
+  }
```
[the remark this answers is not in the document]
Comment (0618) [reply packages/amplify-cli/src/init-steps/preInitSetup.ts:17]: Yea, good catch. Because it was renamed from `isHeadless` dfba32eb5086bd4918c3daebe949ca439c49f543

**Review by sobolk — APPROVED** [review]

**Review by awsluja — APPROVED** [review]
